Update item struct

Middlebrooks314 opened the pull request:
## Summary
The `update-item-struct` branch adds the feature for the user to see the additional item details on the item show page. 

Figure 1. Demo from real device. 

https://user-images.githubusercontent.com/43938783/125510125-669a23a6-b902-473a-a0cc-b54369571f28.mov

whaley-ethan committed:
added attributes to item struct

Middlebrooks314 committed:
update ui

Middlebrooks314 committed:
update constant in test

Middlebrooks314 committed:
update ui text wrapping

elijahsullivan commented on GuildedRoseLLC/Details/DetailsViewController.swift:
```diff
@@ -13,5 +15,16 @@ public class DetailViewController: UIViewController {
         }
         
         itemName.text = item?.name
+        sellIn.text = item?.sellIn ??? "None"
+        quality.text = item?.quality ??? "None"
```
I thought the nil coalescing operator in Swift was `??`, what does `???` do?

ericmeyer commented on GuildedRoseLLC/Items/Item.swift:
```diff
@@ -1,10 +1,18 @@
     
     public init(name: String) {
         self.name = name
+        self.sellIn = 0
```
What is the reason for default `sellIn` and `quality` to `0`?

ericmeyer commented on GuildedRoseLLC/Details/DetailsViewController.swift:
```diff
@@ -13,5 +15,16 @@ public class DetailViewController: UIViewController {
+    }
+}
+
+infix operator ???: NilCoalescingPrecedence
```
I'm a bit confused by what's going on here. What is `NilCoalescingPrecedence`? And more generally, what is this code doing?

ericmeyer commented on GuildedRoseLLC/Details/DetailsViewController.swift:
```diff
@@ -13,5 +15,16 @@ public class DetailViewController: UIViewController {
         }
         
         itemName.text = item?.name
+        sellIn.text = item?.sellIn ??? "None"
```
OK. I think I see what is happening. It seems like you've defined your own custom operator of `???`. Is that accurate? What led to the introduction of this change, and what exactly does `???` do?

ericmeyer commented on GuildedRoseLLC/Items/ItemParser.swift:
```diff
@@ -5,7 +5,6 @@ public class ItemParser {
     public static func parse(json: Data) -> [Item] {
         let decoder = JSONDecoder()
-        
         if let itemsList: [Item] = try? decoder.decode ([Item].self, from: json) {
```
Can `[Item].self` be simplified to just `[Item]`?

Middlebrooks314 replied:
It cannot, it is referring to the class

ericmeyer commented on GuildedRoseLLCSpec/DetailsViewControllerSpec.swift:
```diff
@@ -9,14 +9,62 @@ class DetailsViewControllerSpec: QuickSpec {
+                detailsController.sellIn = UILabel()
+                
                 detailsController.setItem = {
                     () in return Item(name: "FooBar")
```
The way this is formatted makes it a bit hard to read. Typically, the `in` is on the same line as the opening curly brace.

I would recommend looking up and following a style guide.

ericmeyer replied:
Also, I'm unsure as to whether or not `()` before the `in` is conventional, so I would double check that detail too.

Middlebrooks314 replied:
changed

ericmeyer commented on GuildedRoseLLCSpec/DetailsViewControllerSpec.swift:
```diff
@@ -9,14 +9,62 @@ class DetailsViewControllerSpec: QuickSpec {
+                detailsController.sellIn = UILabel()
+                
+                detailsController.setItem = {
+                    () in
```
Same comment regarding style here.

Middlebrooks314 replied:
changed

ericmeyer commented on GuildedRoseLLCSpec/DetailsViewControllerSpec.swift:
```diff
@@ -9,14 +9,62 @@ class DetailsViewControllerSpec: QuickSpec {
+                    var item = Item(name: "FooBar")
+                    item.quality = 5
+                    item.sellIn = 12
+                    return item
```
Does this need to return the `item`?

Middlebrooks314 replied:
yes as part of the closure function.

ericmeyer commented on GuildedRoseLLCSpec/DetailsViewControllerSpec.swift:
```diff
@@ -9,14 +9,62 @@ class DetailsViewControllerSpec: QuickSpec {
+                detailsController.sellIn = UILabel()
+                
+                detailsController.setItem = {
+                    () in
```
Same comment regarding style.

Middlebrooks314 replied:
changed

ericmeyer commented on GuildedRoseLLC/Details/DetailsViewController.swift:
```diff
@@ -2,6 +2,8 @@ import UIKit
 
 public class DetailViewController: UIViewController {
   
+    @IBOutlet public var quality: UILabel!
```
Comment about naming. Right now, you have `quality`, `sellIn`, and `itemName`. This is a bit inconsistent, since one has the prefix of `item`, and the other two do not.

Additionally, when just looking at the variable name, there's nothing to differentiate `quality` the `UILabel` from `quality` the `Int`. I would suggest naming it `qualityLabel` or `itemQualityLabel`, so there is no ambiguity.

ericmeyer commented on GuildedRoseLLC/Details/DetailsViewController.swift:
```diff
@@ -2,6 +2,8 @@ import UIKit
+    @IBOutlet public var quality: UILabel!
+    @IBOutlet public var sellIn: UILabel!
     @IBOutlet public var itemName: UILabel!
     public var setItem: (() -> Item)?
```
I know this is old code, but I love having an explicit setter. It makes something like moving towards a protocol easier.

Middlebrooks314 committed:
Refactor to remove Item constructor

This introduces the ItemTestData class to reduce the test dependencies on the the Item constructor

Middlebrooks314 committed:
Remove the NilColescingOperator to simplify the code

Middlebrooks314 committed:
Update label names for clarity
